src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
+				attributes.setAttribute(KAFKA_RAW_DATA, record);
+			}
+		}
+	}
+
+	@Override
+	protected AttributeAccessor getErrorMessageAttributes(Message<?> message) {
```

Comment:
I don't understand how that will help - this mechanism allows additional attributes to be passed to the EMS; not just the OM.